org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/managers/ContentProviderManager.java

```diff
@@ -0,0 +1,235 @@
+/*******************************************************************************
+ * Copyright (c) 2017 David Gileadi and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     David Gileadi - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.jdt.ls.core.internal.managers;
+
+import java.net.URI;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IConfigurationElement;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.jdt.core.IClassFile;
+import org.eclipse.jdt.ls.core.internal.IContentProvider;
+import org.eclipse.jdt.ls.core.internal.IDecompiler;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+import org.eclipse.jdt.ls.core.internal.preferences.PreferenceManager;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+public class ContentProviderManager {
+
+	private static final String EMPTY_CONTENT = "";
+	private static final String EXTENSION_POINT_ID = "org.eclipse.jdt.ls.core.contentProvider";
+	private static final String CLASS = "class";
+	private static final String ID = "id";
+	private static final String PRIORITY = "priority";
+	private static final String URI_PATTERN = "uriPattern";
+	private static final String CACHEABLE = "cacheable";
+	private static final int DEFAULT_PRIORITY = 500;
+	private static final Pattern DEFAULT_URI_PATTERN = Pattern.compile("jdt://contents/.*\\.class.*");
+	private static final int MAX_CACHE_SIZE = 20;
+
+	private final PreferenceManager preferenceManager;
+
+	private static Cache<String, String> cachedContent = CacheBuilder.newBuilder().maximumSize(MAX_CACHE_SIZE).build();

```

Comment:
I'd rather avoid keeping a static cache instance. You get pretty much the same results if a single instance of ContentProviderManager is shared accross all classes. The reason static cache is bad: it introduces side effects when running unit tests, the cache would be hot, depending on the order of test runs causing potentially hard to debug issues.

You can store the ContentProviderManager instance in JavaLanguageServerPlugin along the rest of the other global "managers". (I don't quite like that pattern and would much rather prefer a dependency injection mechanism, but that's another debate)

Reply:
I'm not overly familiar with Guava's cache. How does it behave during concurrent accesses?

Reply:
[Their docs](https://github.com/google/guava/wiki/CachesExplained) describe it as "similar to `ConcurrentMap`" but with the addition of features like eviction. The docs describe using locks to manage concurrent access.